### PR TITLE
Another tweak to cong!

### PR DESCRIPTION
**Pull request by uncle-betty**
Hello there.

I do a lot of equational reasoning with `_/_` and `_%_`. This means `NonZero` instance arguments on both sides of my equalities. Sometimes, when `cong!` runs, the metas for the instance arguments are already resolved. But sometimes they aren't. Or one of them is, but the other isn't. It depends on how things fall into place during type-checking.

When there are such unresolved metas, `cong!` fails. For example, even an identical instance argument gets a different meta on the LHS vs. the RHS. So, anti-unification considers them a difference and places a phi. Same for an unresolved meta on one side and a resolved meta on the other.

The solution that I would like to propose is to wait for the meta variables to be resolved, before doing anti-unification. However, checking the goal for the existence of metas comes with a cost: the goal needs to be traversed, which impacts `cong!` performance.

This proposal thus only checks for metas once unification failed. The idea is to avoid burdening other people with the meta check, who might not use `_/_` and `_%_` as much as I do, and who thus might not have this problem.

In addition to the meta check, this proposal tries to make the error message issued by a failing `cong!` more meaningful to the user. `cong!` now tells them which term it tried, so that the user can troubleshoot the situation. Consider the following:
```
lemma : ∀ m n o → m + n + o ≡ n + m + o
lemma m n o = begin-equality
  m + n + o ≡⟨ cong! (+-comm m n) ⟩
  n + m + o ∎
  where open ≤-Reasoning
```
This now yields the following error:
```
cong! failed, tried:
cong (λ ϕ → ϕ + ϕ + o) (+-comm m n)
```
Previously the error was the following, which I've always found confusing:
```
m != m + n of type ℕ
```

### CODE REVIEW

**Review by JacquesCarette — APPROVED** [review]
Seems very sensible to me - but I'm also seriously not up-to-date on Agda's reflection, so my approval should be taken with a big lump of salt.

File: src/Tactic/Cong.agda [hunk 1]
```diff
@@ -132,6 +146,21 @@ private
...
+
+      open Traversal applicative
+
+      actions : Actions
+      actions = record defaultActions { onMeta = λ _ x → x ∷ [] }
+
+    blockOnMetas : Term → TC ⊤
+    blockOnMetas t with traverseTerm actions (0 , []) t
+    ... | []         = pure tt
+    ... | xs@(_ ∷ _) = blockTC (blockerAll (List.map blockerMeta xs))
```
Comment (MatthewDaggitt) [inline src/Tactic/Cong.agda:162]: This feels like this (or part of it) could be extracted to a utility method in `Reflection.AST.Traversal` itself?

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Otherwise looks great, and the reasoning seems sound to me! Can you add some test cases to the README file to document the behaviour and guard against regressions? Also could you add a CHANGELOG entry describing the enhancement?

File: src/Tactic/Cong.agda [hunk 2]
```diff
@@ -102,6 +106,16 @@ private
...
   notEqualityError goal = typeError (strErr "Cannot rewrite a goal that is not equality: " ∷ termErr goal ∷ [])
 
+  unificationError : ∀ {A : Set} → TC Term → TC Term → TC A
+  unificationError term symTerm = do
+    term' ← term
+    symTerm' ← symTerm
+    -- Don't show the same term twice.
+    let symErr = case term' Term.≟ symTerm' of λ where
+      (yes _) → []
+      (no _) → strErr "\n" ∷ termErr symTerm' ∷ []
```
[the remark this answers is not in the document]
Comment (gallais) [reply src/Tactic/Cong.agda:116]: Yes you do.

File: src/Tactic/Cong.agda [hunk 3]
```diff
@@ -236,12 +265,17 @@ macro
...
+      let makeTerm = λ lhs rhs → `cong eqGoal (antiUnify 0 lhs rhs) x≡y
       let lhs = EqualityGoal.lhs eqGoal
       let rhs = EqualityGoal.rhs eqGoal
-      -- When using ⌞_⌟ with ≡⟨ ... ⟨, (uni lhs rhs) fails and
-      -- (uni rhs lhs) succeeds.
-      catchTC (uni lhs rhs) (uni rhs lhs)
+      let term = makeTerm lhs rhs
+      let symTerm = makeTerm rhs lhs
+      let uni = _>>= flip unify hole
+      catchTC
```
[the remark this answers is not in the document]
Comment (uncle-betty) [reply src/Tactic/Cong.agda:274]: Yes! And it should.

File: src/Tactic/Cong.agda (hunk 1 above)
Comment (uncle-betty) [reply src/Tactic/Cong.agda:162]: I like the idea! Other tactics might want to block on metas, too. But please help me understand: The `Traversal` module takes an applicative as an argument. Wouldn't a user then have to provide a "dummy" applicative in order to use `blockOnMetas`? (Which would ignore the user-provided applicative.)

Comment (uncle-betty) [reply src/Tactic/Cong.agda:162]: What I'm trying to say: Should there be a new module, `Reflection.AST.Utilities`, that would be home to `blockOnMetas`?

**Comment by uncle-betty** [comment]
Thank you for your reviews!

I think I might need some guidance regarding the `blockOnMetas` (see my response to the comment), but I made the other changes.

The idea behind the `test₂` unit test, which may not be completely obvious:

* Anecdotally, I seem to have encountered issues mostly with reasoning chains that were a little longer. Hence the redundant steps.
* It tests explicitly given instances in addition to instances from instance arguments.

File: doc/README/Tactic/Cong.agda [hunk 4]
```diff
@@ -161,3 +162,34 @@ module EquationalReasoningTests where
     ≤⟨ n≤1+n _ ⟩
       suc (suc (n + n))
     ∎
+
+module MetaTests where
```
Comment (MatthewDaggitt) [inline doc/README/Tactic/Cong.agda:166]: Should be an empty line here.

File: src/Tactic/Cong.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Tactic/Cong.agda:162]: Ah sorry, I missed that `Traversal` is parameterised by an applicative. Yes a new module `Reflection.TCM.Utilities` would be good. (Note I've had a change of heart from `AST` to `TCM`!)

File: doc/README/Tactic/Cong.agda (hunk 4 above)
Comment (uncle-betty) [reply doc/README/Tactic/Cong.agda:166]: Oh, right. Yes.

File: src/Tactic/Cong.agda (hunk 1 above)
Comment (uncle-betty) [reply src/Tactic/Cong.agda]: Alright. Great. I moved it.

**Comment by uncle-betty** [comment]
Thanks for taking another look. I made the changes.

**Comment by jamesmckinna** [comment]
Coming to this PR quite late...
... if you find yourself doing a lot of equational reasoning with `_/_` and `_%_` then is there a higher level (as opposed to low-level `cong` reasoning) collection of `DivMod` abstractions which might streamline your work?

**Comment by uncle-betty** [comment]
Excellent question. I don't know yet. I've only started toying around.

I'm curious about modeling machine integers in Agda: convert integers modulo 2^64 to `Vec Bool 64` and vice versa, prove that division by 2 of an integer corresponds to a right shift of the corresponding bit vector - this kind of thing.

Yes, there may be a much better way than what I'm currently doing. But I still think that there's benefit in making `cong!` deal properly with meta variables in goals.

**Comment by jamesmckinna** [comment]
@uncle-betty you wrote: 
> I'm curious about modeling machine integers in Agda: convert integers modulo 2^64 to `Vec Bool 64` and vice versa, prove that division by 2 of an integer corresponds to a right shift of the corresponding bit vector - this kind of thing.

Makes good sense, but then I would (be tempted to) avoid `_/_` and `_%_` in favour of using the `Data.Nat.Base.⌊_/2⌋` and `Data.Nat.Base.⌈_/2⌉` functions and their properties, and perhaps consider iterates of them for `m-`ary shifts? If you're only considering base-`2^n` arithmetic, I think full 'div-mod' is overkill?

> Yes, there may be a much better way than what I'm currently doing. But I still think that there's benefit in making `cong!` deal properly with meta variables in goals.

Absolutely fair point.

**Comment by uncle-betty** [comment]
Oh, nice! Thanks, @jamesmckinna, for pointing me to these functions. I appreciate the guidance.

**Review by MatthewDaggitt — APPROVED** [review]

**Comment by MatthewDaggitt** [comment]
Apologies for the delay. Merging in now. Thanks for another great PR!

**Comment by uncle-betty** [comment]
Oh, that's exciting! Great! Thank you for the merge and for your assistance.

**Pull request merged**